test/indexing.jl

```diff
@@ -0,0 +1,171 @@
+                        label=["x", "y"],
+                    ),
+                ],
+            ])...
+        )
+
+        grp1a = ds[(:group1, :a)]
+        @test isa(grp1a, KeyedArray)
+
+        grp2b = ds[(:group2, :b)]
+        @test isa(grp2b, KeyedArray)
+
+        @test_throws KeyError ds[(:group1, :)]
+        @test_throws KeyError ds[:group1]
+        @test_throws KeyError ds[:a]
+    end
+end
+
+@testset "setindex!" begin
+    ds = KeyedDataset(
+        :a => KeyedArray(
+            allowmissing(rand(4, 3, 2));
+            time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+            loc=1:3,
+            obj=[:a, :b],
+        ),
+        :b => KeyedArray(
+            rand(4, 2);
+            time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+            label=["x", "y"],
+        ),
+    )
+
+    init_constraints = deepcopy(ds.constraints)
+    @test length(init_constraints) == 4
+
+    ds[:c] = KeyedArray(
+        rand(4, 2);
+        time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+        label=["x", "y"],
+    )
+    @test ds.constraints == init_constraints
+
+    ds[:c] = KeyedArray(rand(5); id=100:100:500)
+    @test ds.constraints != init_constraints
+    @test length(ds.constraints) == 5
+
+    # Test validation error if we try to add a component with a mismatched dimension
+    # relative to our existing dimensions / constraints
+    @test_throws ArgumentError setindex!(ds, KeyedArray(rand(2); id=[1, 2]), :d)
+end
+
+@testset "lookup" begin
+    ds = KeyedDataset(
+        flatten([
+            :group1 => [
+                :a => KeyedArray(
+                    allowmissing(rand(4, 3, 2));
+                    time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                    loc=1:3,
+                    obj=[:a, :b],
+                ),
+                :b => KeyedArray(
+                    rand(4, 2);
+                    time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                    label=["x", "y"],
+                )
+            ],
+            :group2 => [
+                :a => KeyedArray(
+                    rand(4, 3, 2) .+ 1.0;
+                    time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                    loc=1:3,
+                    obj=[:a, :b],
+                ),
+                :b => KeyedArray(
+                    rand(4, 2) .+ 1.0;
+                    time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                    label=["x", "y"],
+                )
+            ]
+        ])...
+    )
+
+    a = ds(:_, :a)
+    @test isa(a, KeyedDataset)
+    @test issetequal([(:group1, :a), (:group2, :a)], keys(a.data))
+
+    grp1 = ds(:group1, :__)
+    @test isa(grp1, KeyedDataset)
+    @test issetequal([(:group1, :a), (:group1, :b)], keys(grp1.data))
+
+    @test_throws KeyError ds(:group1)
+    @test_throws KeyError ds(:a)
+
+    # Test custom filter
```

Comment:
Not really, but it's easy to support. I guess this kind of came out of Sam's suggestion of using a regex instead of the `Pattern` syntax. I think that'd be more expensive, but if you have some custom naming convention, then you could use a regex instead of a `Pattern` to find it.